proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1155 @@
+   variable, later used used in the type signature of ``g``.
+
+   The ``b`` is bound implicitly in this example, assuming there's no top-level
+   definition of ``b``. To make it clean, one can use an explicit ``forall``::
+
+     ... where
+             g :: forall b. b -> x
+             g _ = y
+
+   This is similar to the situation with ``ScopedTypeVariables``, where we also
+   cannot assume that all lowercase variables in a signature are free.
+
+Effect and Interactions
+-----------------------
+
+* Visible ``forall`` becomes available in types of terms, making them more
+  similar to types of types. There remains a discrepancy that ``forall`` in
+  types of types is actually a relevant quantifier, while the proposed ``forall
+  x ->`` in types of terms is irrelevant. This is to be resolved in the future
+  by making both of them irrelevant.
+
+* The renaming of a visible dependent argument is different than that of a
+  dependent argument with a visibility override. Consider this code::
+
+    f :: forall a.   Tagged a ()
+    g :: forall a -> Tagged a ()
+
+    data T = T
+
+    a = f @T
+    b = g  T
+
+  In ``f @T``, we refer to the type constructor, but in ``g T`` we refer to the
+  data constructor.
+
+  The implementation may offer warning flags to help the user identify such
+  ambiguous occurrences.
+
+* When punned names come from external libraries, there are several workarounds
+  to resolve the ambiguity:
+
+  1. Using ``-XExplicitNamespaces``::
+
+      import Data.Proxy
+      import qualified Data.Proxy (type Proxy) as T
+
+      x = f   Proxy  -- refers to the data constructor
+      y = f T.Proxy  -- refers to the type constructor
+
+  2. Using a type synonym::
+
+      import Data.Proxy
+
+      type TProxy = Proxy
+
+      x = f  Proxy  -- refers to the data constructor
+      y = f TProxy  -- refers to the type constructor
+
+  3. Using the ``type`` herald::
+
+      import Data.Proxy
+
+      x = f Proxy
+      y = f (type Proxy)
+
+* Identifiers bound to terms are
+  not promoted. Consider this well-typed program::
+
+    f :: forall a.   Tagged a ()
+    g :: forall a -> Tagged a ()
+
+    a = f @(Just True)  -- ok
+    b = g  (Just True)  -- ok
+
+  If we factor out ``Just True`` into a type synonym, it continues to work::
+
+    type X :: Maybe Bool
+    type X = Just True
+
+    a = f @X  -- still OK
+    b = g  X  -- still OK
+
+  However, if we bind it to a term-level variable, the example becomes
+  ill-typed::
+
+    x :: Maybe Bool
+    x = Just True
+
+    a = f @x  -- not currently valid
+    b = g  x  -- not valid under the proposal
+
+  This is because we retain the distinction between terms and types. This
+  proposal is a step towards dependent types, but it does not go all the way.
+  Accepting the program above is left as future work.
+
+Costs and Drawbacks
+-------------------
+
+This is one more feature to implement and maintain.
```

Comment:
Do you have an estimate on how big a feature this is, in terms of implementation?